crates/cadenza-eval/src/eval.rs

```diff
@@ -0,0 +1,458 @@
+//! Tree-walk evaluator for the Cadenza language.
+//!
+//! The evaluator interprets the AST produced by cadenza-syntax.
+//! It handles:
+//! - Literals (integers, floats, strings)
+//! - Identifiers (variable lookup)
+//! - Lists/vectors
+//! - Function application
+//! - Macro expansion
+
+use crate::compiler::Compiler;
+use crate::env::Env;
+use crate::error::{Error, Result};
+use crate::interner::{InternedId, Interner};
+use crate::value::Value;
+use cadenza_syntax::ast::{Apply, Attr, Expr, Ident, Literal, LiteralValue, Root, Synthetic};
+
+/// Evaluates a complete source file (Root node).
+///
+/// Each top-level expression is evaluated in order. The results are
+/// collected into a vector, though most top-level expressions will
+/// return `Value::Nil` as side effects on the `Compiler` are the
+/// primary purpose.
+pub fn eval(root: &Root, interner: &mut Interner, env: &mut Env, compiler: &mut Compiler) -> Result<Vec<Value>> {
+    let mut results = Vec::new();
+    for expr in root.items() {
+        let value = eval_expr(&expr, interner, env, compiler)?;
+        results.push(value);
+    }
+    Ok(results)
+}
+
+/// Evaluates a single expression.
+pub fn eval_expr(
+    expr: &Expr,
+    interner: &mut Interner,
+    env: &mut Env,
+    compiler: &mut Compiler,
+) -> Result<Value> {
+    match expr {
+        Expr::Literal(lit) => eval_literal(lit),
+        Expr::Ident(ident) => eval_ident(ident, interner, env, compiler),
+        Expr::Apply(apply) => eval_apply(apply, interner, env, compiler),
+        Expr::Attr(attr) => eval_attr(attr, interner, env, compiler),
+        Expr::Op(op) => {
+            // Operators as values (for higher-order usage)
+            let text = op.syntax().text().to_string();
```

Comment:
I really want to avoid allocating when checking the intern map. We'll probably have to use the hashbrown crate directly where you can get the bucket for a borrowed key and potentially clone and insert it if it doesn't exist.